collect/concurrency_limiter.go

```diff
@@ -0,0 +1,36 @@
+package collect
+
+type concurrencyLimiter struct {
+	limit chan struct{}
+	done  chan struct{}
+}
+
+func newConcurrencyLimiter(n int) *concurrencyLimiter {
+	return &concurrencyLimiter{
+		limit: make(chan struct{}, n),
+		done:  make(chan struct{}),
+	}
+}
+
+func (l *concurrencyLimiter) Go(fn func()) {
+
+	select {
+	case l.limit <- struct{}{}:
+	case <-l.done:
+		return
```

Comment:
Do we want to block here instead of returning? Otherwise we're just going to accept work and not respond to it. Also, why use this instead of the library you had yesterday?

Reply:
The done channel is only closed when we are done with all the work. This is to prevent new work being started. Maybe we can return a boolean to indicate whether a work is started?

I don't think I have used other library for this. Do you mind to clarify which library you are referring to?

Reply:
I meant errgroup.Group, which has a [SetLimit](https://pkg.go.dev/golang.org/x/sync/errgroup#Group.SetLimit) method.